docs/recipes/recipes.ipynb

```diff
@@ -103,6 +103,276 @@
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "aa1df625-9f6a-44d3-9169-1c79ff8e647f",
+   "metadata": {
+    "jp-MarkdownHeadingCollapsed": true
+   },
+   "source": [
+    "### Setup\n",
+    "\n",
+    "Lets look at a situation where we have some \"raw\" data files and the workflow consists of three steps\n",
+    "  * loading the raw data\n",
+    "  * cleaning the raw data\n",
+    "  * computing a sum of the cleaned data."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "bb0ecea3-5b0d-44da-a363-2a0e861b0235",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from typing import NewType\n",
+    "\n",
+    "Filename = NewType('Filename', str)\n",
+    "RawData = NewType('RawData', list)\n",
+    "CleanData = NewType('CleanData', list)\n",
+    "Result = NewType('Result', list)\n",
+    "\n",
+    "filesystem = {'raw.txt': list(map(str, range(10)))}\n",
+    "\n",
+    "def load(filename: Filename) -> RawData:\n",
+    "    \"\"\"Load the data from the filename.\"\"\"\n",
+    "    data = filesystem[filename]\n",
+    "    return RawData(data)\n",
+    "\n",
+    "def clean(raw_data: RawData) -> CleanData:\n",
+    "    \"\"\"Clean the data, convert from str.\"\"\"\n",
+    "    return CleanData(list(map(float, raw_data)))\n",
+    "\n",
+    "def process(clean_data: CleanData) -> Result:\n",
+    "    \"\"\"Compute the sum of the clean data.\"\"\"\n",
+    "    return Result(sum(clean_data))\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "1a4d59ab-4022-4eef-8566-d1b37f9cea7f",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import sciline\n",
+    "\n",
+    "pipeline = sciline.Pipeline(\n",
+    "    [load, clean, process,],\n",
+    "    params={ Filename: 'raw.txt', })\n",
+    "pipeline"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a7870758-aa28-497d-adc9-863efe23e463",
+   "metadata": {},
+   "source": [
+    "### Setting intermediate results"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "74ace6e8-2e32-420b-b96d-dc67c8ce4ab7",
+   "metadata": {},
+   "source": [
+    "If we select `Result` the task graph will use the `Filename` input because it needs to read the raw data from the file system:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "837135eb-c858-484e-91b5-b47917aefe57",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "pipeline.get(Result)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "99fe0c69-597b-4384-96de-92b25b6e95cd",
+   "metadata": {},
+   "source": [
+    "But if the cleaned data has already been produced it is unnecessary to \"re-clean\" it, in that case we can proceed directly from the clean data to the compute sum step.\n",
+    "To do this we replace the `CleanData` provider with the data that was loaded and cleaned:"
```

Comment:
This is a bit confusing, because the *computation* of `CleanData` comes after the above line, where you say we replace the provider. A novice user might thing the call to `compute` is part of the replacement. How about something like:

Given a pipeline, we may want to compute an intermediate result for inspection:
```python
data = pipeline.compute(CleanData)
```

If later on we wish to compute a result further down the pipeline (derived from `CleanData`), this would cause potentially very costly re-computation of `CleanData`, since Sciline does not perform any caching:
```python
result = pipeline.compute(Result)  # re-computes CleanData
```

To avoid this, we can use `Pipeline.__setitem__` to replace the provider of `CleanData` by the previously computed data:
```python
pipeline[CleanData] = data
result = pipeline.compute(Result)  # uses data as CleanData
```